### ISSUE TITLE
Defunctionalization of type families

### ISSUE THREAD

**Issue by langston-barrett** [GaloisInc/parameterized-utils#105]
[The technique of defunctionalizing type families](https://blog.jle.im/entry/introduction-to-singletons-4.html#title) to support partial application (and especially higher-order type families) has been applied in at least two `parameterized-utils`-adjacent spots: #103, and [the `CtxFuns` module](https://github.com/GaloisInc/asl-translator/blob/feature/ctxfuns/lib/Data/Parameterized/CtxFuns.hs) mentioned in #84. Seems like it's probably worth creating a small module in `parameterized-utils` that at least provides a canonical home for the `Apply` family:
```haskell
data TyFun a b
type a ~> b = TyFun a b -> Type
type family Apply (f :: a ~> b) (x :: a) :: b
```

**Comment by RyanGlScott** [comment]
Alternatively, we could just depend on the [`singletons`](http://hackage.haskell.org/package/singletons-3.0/docs/Data-Singletons.html#t:Apply) library, which provides this in `Data.Singletons`. `singletons-3.0` and later have no external dependencies, so it's a pretty lightweight dependency.

**Comment by travitch** [comment]
Yeah, there is also a version in semmc somewhere, I think. It would be nice to have just one. I'm open to singletons core now that it has slimmed down.

**Comment by langston-barrett** [comment]
Let's close this and just recommend `singletons`.
